File: Yandex/Practicum/Algorightms_and_Data_Structures/8_sprint/D._Общий_префикс/solution.py
import sys


def solution() -> None:
    """
    Time Complexity O(L)
    Memory Complexity O(M)
    """
    n: int = int(sys.stdin.readline().rstrip())
    pattern: str = ''
    p_n: int = 0
    for i in range(n):
        word: str = sys.stdin.readline().rstrip()
        if not i:
            pattern = word
            p_n = len(pattern)
        else:
            idx: int = 0
            while idx < min(p_n, len(word)):
                if pattern[idx] == word[idx]:
                    idx += 1
                else:
                    break
            p_n = idx
    sys.stdout.write('{}\n'.format(p_n))


if __name__ == '__main__':
    solution()

File: Yandex/Practicum/Algorightms_and_Data_Structures/8_sprint/D._Общий_префикс/solution.cpp
#include <iostream>
#include <string>


void solution() {
    /*
    Time Complexity O(L)
    Memory Complexity O(M)
    */
    int n;
    std::scanf("%d", &n);
    std::string pattern;
    int pN = 0;
    for (int i = 0; i < n; ++i) {
        std::string word;
        std::cin >> word;
        if (!i) {
            pattern = word;
            pN = pattern.size();
        }
        else {
            int idx = 0;
            while (idx < std::min(pN, (int)word.size())) {
                if (pattern[idx] == word[idx]) ++idx;
                else break;
            }
            pN = idx;
        }
    }
    std::cout << pN << "\n";
}


int main() {
    solution();
}